cpp/src/interop/to_arrow_device.cu

```diff
@@ -0,0 +1,590 @@
+{
+  return ArrowSchemaSetTypeDecimal(out,
+                                   NANOARROW_TYPE_DECIMAL128,
+                                   cudf::detail::max_precision<DeviceType>(),
+                                   -input.type().scale());
+}
+
+template <>
+int dispatch_to_arrow_type::operator()<numeric::decimal32>(column_view input,
+                                                           cudf::type_id,
+                                                           column_metadata const&,
+                                                           struct ArrowSchema* out)
+{
+  using DeviceType = int32_t;
+  return decimals_to_arrow<DeviceType>(input, out);
+}
+
+template <>
+int dispatch_to_arrow_type::operator()<numeric::decimal64>(column_view input,
+                                                           cudf::type_id,
+                                                           column_metadata const&,
+                                                           struct ArrowSchema* out)
+{
+  using DeviceType = int64_t;
+  return decimals_to_arrow<DeviceType>(input, out);
+}
+
+template <>
+int dispatch_to_arrow_type::operator()<numeric::decimal128>(column_view input,
+                                                            cudf::type_id,
+                                                            column_metadata const&,
+                                                            struct ArrowSchema* out)
+{
+  using DeviceType = __int128_t;
+  return decimals_to_arrow<DeviceType>(input, out);
+}
+
+template <>
+int dispatch_to_arrow_type::operator()<bool>(column_view input,
+                                             cudf::type_id,
+                                             column_metadata const&,
+                                             struct ArrowSchema* out)
+{
+  return ArrowSchemaSetType(out, NANOARROW_TYPE_BOOL);
+}
+
+template <>
+int dispatch_to_arrow_type::operator()<cudf::string_view>(column_view input,
+                                                          cudf::type_id,
+                                                          column_metadata const&,
+                                                          struct ArrowSchema* out)
+{
+  return ArrowSchemaSetType(out, NANOARROW_TYPE_STRING);
+}
+
+template <>
+int dispatch_to_arrow_type::operator()<cudf::list_view>(column_view input,
+                                                        cudf::type_id,
+                                                        column_metadata const& metadata,
+                                                        struct ArrowSchema* out);
+
+template <>
+int dispatch_to_arrow_type::operator()<cudf::dictionary32>(column_view input,
+                                                           cudf::type_id,
+                                                           column_metadata const& metadata,
+                                                           struct ArrowSchema* out);
```

Comment:
Because I recursively use the `dispatch_to_arrow_type` inside of the struct specialization, I needed to forward declare these or else I get a "specialization declared after use" compile error.

Reply:
Ah got it I forgot about the recursion, you're right.